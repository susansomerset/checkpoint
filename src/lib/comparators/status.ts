export const STATUS_PRIORITY = ['Missing','Submitted (Late)','Submitted','Graded'] as const;
const rank = Object.fromEntries(STATUS_PRIORITY.map((s,i)=>[s,i]));

export function compareStatus(a: string, b: string) {
  return (rank[a] ?? 999) - (rank[b] ?? 999);
}
