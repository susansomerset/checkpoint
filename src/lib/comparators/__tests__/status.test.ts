import { STATUS_PRIORITY, compareStatus } from '../status';

describe('STATUS_PRIORITY', () => {
  test('maintains exact order and never changes', () => {
    // Snapshot test to catch any accidental reordering
    expect(STATUS_PRIORITY).toMatchSnapshot();
    
    // Explicit order verification
    expect(STATUS_PRIORITY).toEqual([
      'Missing',
      'Submitted (Late)', 
      'Submitted',
      'Graded'
    ]);
  });

  test('compareStatus respects priority order', () => {
    // Test all combinations to ensure proper ordering
    expect(compareStatus('Missing', 'Submitted (Late)')).toBeLessThan(0);
    expect(compareStatus('Submitted (Late)', 'Submitted')).toBeLessThan(0);
    expect(compareStatus('Submitted', 'Graded')).toBeLessThan(0);
    
    // Test reverse order
    expect(compareStatus('Graded', 'Missing')).toBeGreaterThan(0);
    expect(compareStatus('Submitted', 'Submitted (Late)')).toBeGreaterThan(0);
    
    // Test equal status
    expect(compareStatus('Missing', 'Missing')).toBe(0);
    
    // Test unknown status (should sort last)
    expect(compareStatus('Unknown', 'Missing')).toBeGreaterThan(0);
    expect(compareStatus('Graded', 'Unknown')).toBeLessThan(0);
  });
});
