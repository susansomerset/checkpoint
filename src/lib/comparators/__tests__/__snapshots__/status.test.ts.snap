// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`STATUS_PRIORITY maintains exact order and never changes 1`] = `
[
  "Missing",
  "Submitted (Late)",
  "Submitted",
  "Graded",
]
`;
